File: core/parser/clash.py
"""
Парсер конфигураций формата Clash / Mihomo (YAML).
"""

from typing import Any, Dict, List
import yaml

from core.config.models import ProfileConfig, ProtocolType
from core.parser.base import BaseParser, ParseResult
from utils.logger import log_call


class ClashParser(BaseParser):
    """Парсер для YAML-конфигураций Clash и Mihomo."""

    format_name: str = "clash"

    # Маппинг типов Clash прокси на наш ProtocolType
    PROTOCOL_MAP = {
        "ss": ProtocolType.SHADOWSOCKS,
        "shadowsocks": ProtocolType.SHADOWSOCKS,
        "vmess": ProtocolType.VMESS,
        "vless": ProtocolType.VLESS,
        "trojan": ProtocolType.TROJAN,
        "hysteria2": ProtocolType.HYSTERIA2,
        "hy2": ProtocolType.HYSTERIA2,
        "wireguard": ProtocolType.WIREGUARD,
        "tuic": ProtocolType.TUIC,
        "ssh": ProtocolType.SSH,
        "direct": ProtocolType.DIRECT,
        "reject": ProtocolType.BLOCK,
    }

    @log_call
    def parse(self, content: str) -> ParseResult:
        """
        Разбирает YAML-конфигурацию Clash.

        Args:
            content: Текст YAML-файла.

        Returns:
            ParseResult: Результат распарсенной конфигурации.
        """
        result = ParseResult(format_name=self.format_name)

        try:
            data = yaml.safe_load(content)
        except yaml.YAMLError as exc:
            result.errors.append(f"Ошибка синтаксиса YAML Clash: {exc}")
            return result

        if not isinstance(data, dict):
            result.errors.append("Корневой элемент Clash YAML должен быть словарем.")
            return result

        result.raw_config = data
        proxies = data.get("proxies", [])

        if not isinstance(proxies, list):
            result.errors.append("Секция 'proxies' должна быть списком.")
            return result

        for item in proxies:
            if not isinstance(item, dict):
                continue

            name = item.get("name", "Clash Proxy")
            raw_type = str(item.get("type", "")).lower()
            server = item.get("server", "127.0.0.1")
            port = int(item.get("port", 443))

            protocol = self.PROTOCOL_MAP.get(raw_type, ProtocolType.VLESS)

            # Переносим специфичные настройки прокси
            settings = {k: v for k, v in item.items() if k not in ("name", "type", "server", "port")}

            profile = ProfileConfig(
                name=name,
                protocol=protocol,
                server=server,
                server_port=port,
                settings=settings,
                tag=name,
            )
            result.profiles.append(profile)

        return result

File: core_manager/src/parser/clash.rs
//! Парсер YAML конфигураций Clash / Mihomo на Rust.

use crate::models::{ProfileConfig, ProtocolType};
use crate::parser::{ConfigParser, ParseResult};

pub struct ClashParser;

impl ConfigParser for ClashParser {
    fn format_name(&self) -> &'static str {
        "clash"
    }

    fn parse(&self, content: &str) -> ParseResult {
        let mut result = ParseResult::new(self.format_name());

        let yaml_val: serde_yaml::Value = match serde_yaml::from_str(content) {
            Ok(v) => v,
            Err(err) => {
                result.errors.push(format!("Ошибка синтаксиса YAML Clash: {}", err));
                return result;
            }
        };

        if let Ok(json_val) = serde_json::to_value(&yaml_val) {
            result.raw_config = Some(json_val.clone());
            if let Some(proxies) = json_val.get("proxies").and_then(|p| p.as_array()) {
                for item in proxies {
                    let name = item.get("name").and_then(|n| n.as_str()).unwrap_or("Clash Proxy");
                    let raw_type = item.get("type").and_then(|t| t.as_str()).unwrap_or("vless").to_lowercase();
                    let server = item.get("server").and_then(|s| s.as_str()).unwrap_or("127.0.0.1");
                    let port = item.get("port").and_then(|p| p.as_u64()).unwrap_or(443) as u16;

                    let protocol = match raw_type.as_str() {
                        "ss" | "shadowsocks" => ProtocolType::Shadowsocks,
                        "vmess" => ProtocolType::Vmess,
                        "vless" => ProtocolType::Vless,
                        "trojan" => ProtocolType::Trojan,
                        "hysteria2" | "hy2" => ProtocolType::Hysteria2,
                        "wireguard" => ProtocolType::Wireguard,
                        "tuic" => ProtocolType::Tuic,
                        "ssh" => ProtocolType::Ssh,
                        "direct" => ProtocolType::Direct,
                        "reject" => ProtocolType::Block,
                        _ => ProtocolType::Vless,
                    };

                    let mut p = ProfileConfig::new(name, protocol, server, port);
                    p.settings = item.clone();
                    result.profiles.push(p);
                }
            }
        }

        result
    }
}
